README.md
# Embedded SDK
Provides an example of how to build embedded system simulators with CMake for
hosted operating systems and compilers.

[Changelog](#CHANGES.md)

## Supported Platforms
1. Linux/GCC
2. Linux/Clang
3. Windows/MSVC

## Requirements
**All Platforms**
- Compiler
- CMake
- Git

**Linux**
- `libpcap`
- `pkgconfig`

**Windows**
- `winpcap`

## Selected Libraries
- FreeRTOS
- FreeRTOS+TCP
- FreeRTOS+coreMQTT
- littlefs
- mbedtls

## Default Configuration
By default, Embedded SDK selects some reasonable defaults for desktop simulators.

Additional configuration instructions are under construction.

## Examples
**HelloTask**
2 Tasks periodically printing their names to stdout.

**TcpEchoClientServer**
A client that periodically sends "Hello From Client" to a server.


CHANGES.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.0.0] - 2020-03-28
### Added
- FreeRTOS Kernel version detection support.
- Linux TCP prototype


examples/TcpEchoClientServer/main.c
// std
#include <stdbool.h>
#include <stddef.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>

// 3rd party
#include <FreeRTOS.h>
#include <FreeRTOS_IP.h>
#include <FreeRTOS_Sockets.h>
#include <queue.h>
#include <semphr.h>
#include <task.h>

#define STRINGIFY(expr)  #expr
#define XSTRINGIFY(expr) STRINGIFY(expr)

/* Default MAC address configuration.  The demo creates a virtual network
 * connection that uses this MAC address by accessing the raw Ethernet data
 * to and from a real network connection on the host PC.  See the
 * configNETWORK_INTERFACE_TO_USE definition for information on how to configure
 * the real network connection to use. */
uint8_t ucMACAddress[6] = { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 };

typedef struct Task {
  TaskHandle_t Handle;
  StaticTask_t ControlBlock;
  StackType_t  Stack[configMINIMAL_STACK_SIZE];
} Task;

static Task             NetTask;
static Task             ServerTask;
static const TickType_t DEFAULT_TIMEOUT = pdMS_TO_MIN_TICKS(2000);

bool CheckSocketResult(BaseType_t result, const char* name)
{
  bool ok = true;
  if (result == 0) { printf("%s succeeded\n", name); }
  else
  {
    ok = false;
    printf("%s failed with code %ld\n", name, result);
  }
  return ok;
}

void NetTask_Run(void* parameters)
{
  (void)parameters;
  BaseType_t               socket_result  = 0;
  struct freertos_sockaddr remote_address = {
    .sin_port = FreeRTOS_htons((uint16_t)10000),
    .sin_addr = FreeRTOS_inet_addr_quick(127, 0, 0, 1),
  };

  Socket_t socket = FreeRTOS_socket(FREERTOS_AF_INET, FREERTOS_SOCK_STREAM, FREERTOS_IPPROTO_TCP);
  configASSERT(socket != FREERTOS_INVALID_SOCKET);
  socket_result = FreeRTOS_setsockopt(socket, 0, FREERTOS_SO_RCVTIMEO, &DEFAULT_TIMEOUT, sizeof(DEFAULT_TIMEOUT));
  CheckSocketResult(socket_result, "[Client] sockopt RCVTIMEO");
  socket_result = FreeRTOS_setsockopt(socket, 0, FREERTOS_SO_SNDTIMEO, &DEFAULT_TIMEOUT, sizeof(DEFAULT_TIMEOUT));
  CheckSocketResult(socket_result, "[Client] sockopt SNDTIMEO");

  static const char tx[] = "Hello From Client!";

  do
  {
    socket_result = FreeRTOS_connect(socket, &remote_address, sizeof(remote_address));
  } while (!CheckSocketResult(socket_result, "[Client] connect"));

  while (true)
  {
    size_t total_sent = 0;
    while (total_sent < sizeof(tx))
    {
      size_t     tx_size = sizeof(tx) - total_sent;
      BaseType_t sent    = FreeRTOS_send(socket, &tx[total_sent], tx_size, 0);

      if (sent >= 0) { total_sent += sent; }
      else
      {
        CheckSocketResult(sent, "[Client] send");
      }
    }
    vTaskDelay(DEFAULT_TIMEOUT);
  }

  FreeRTOS_shutdown(socket, FREERTOS_SHUT_RDWR);
  FreeRTOS_closesocket(socket);
  vTaskDelete(NULL);
}

void ServerTask_Run(void* parameters)
{
  (void)parameters;
  static const TickType_t  WAIT_FOREVER    = portMAX_DELAY;
  static const BaseType_t  MAX_CONNECTIONS = 5;
  struct freertos_sockaddr bind_address    = {
    .sin_port = FreeRTOS_htons((uint16_t)10000),
  };
  BaseType_t socket_result = 0;

  Socket_t listen_socket = FreeRTOS_socket(FREERTOS_AF_INET, FREERTOS_SOCK_STREAM, FREERTOS_IPPROTO_TCP);
  configASSERT(listen_socket != FREERTOS_INVALID_SOCKET);
  socket_result = FreeRTOS_setsockopt(listen_socket, 0, FREERTOS_SO_RCVTIMEO, &WAIT_FOREVER, sizeof(WAIT_FOREVER));
  CheckSocketResult(socket_result, "[Server] sockopt RCVTIMEO");
  socket_result = FreeRTOS_bind(listen_socket, &bind_address, sizeof(bind_address));
  CheckSocketResult(socket_result, "[Server] bind");
  socket_result = FreeRTOS_listen(listen_socket, MAX_CONNECTIONS);
  CheckSocketResult(socket_result, "[Server] listen");

  struct freertos_sockaddr client;
  socklen_t                client_size      = sizeof(client);
  Socket_t                 connected_socket = FreeRTOS_accept(listen_socket, &client, &client_size);
  configASSERT(connected_socket != FREERTOS_INVALID_SOCKET);

  static char rx[32];
  memset(rx, 0, sizeof(rx));

  while (true)
  {
    BaseType_t rx_size = FreeRTOS_recv(connected_socket, rx, sizeof(rx), 0);
    if (rx_size > 0) { printf("[Server] Received: %s\n", rx); }
    else if (rx_size < 0)
    {
      CheckSocketResult(rx_size, "[Server] recv");
    }
  }
  FreeRTOS_shutdown(connected_socket, FREERTOS_SHUT_RDWR);
  FreeRTOS_closesocket(connected_socket);
  vTaskDelete(NULL);
}

void CreateTasks()
{
  ServerTask.Handle =
    xTaskCreateStatic(ServerTask_Run, STRINGIFY(ServerTask_Run), sizeof(ServerTask.Stack) / sizeof(*ServerTask.Stack),
                      STRINGIFY(ServerTask_Run), 1, ServerTask.Stack, &ServerTask.ControlBlock);
  NetTask.Handle =
    xTaskCreateStatic(NetTask_Run, STRINGIFY(NetTask_Run), sizeof(NetTask.Stack) / sizeof(*NetTask.Stack),
                      STRINGIFY(NetTask_Run), 1, NetTask.Stack, &NetTask.ControlBlock);
}

int main()
{
  //  static const uint8_t ucIPAddress[4]      = { 10, 10, 10, 200 };
  static const uint8_t ucIPAddress[4] = { 127, 0, 0, 1 };
  static const uint8_t ucNetMask[4]   = { 255, 0, 0, 0 };
  //  static const uint8_t ucGatewayAddress[4] = { 10, 10, 10, 1 };
  static const uint8_t ucGatewayAddress[4] = { 127, 0, 0, 1 };
  /* The following is the address of an OpenDNS server. */
  static const uint8_t ucDNSServerAddress[4] = { 208, 67, 222, 222 };
  FreeRTOS_IPInit(ucIPAddress, ucNetMask, ucGatewayAddress, ucDNSServerAddress, ucMACAddress);
  CreateTasks();
  vTaskStartScheduler();

  return 0;
}


tools/cmake/modules/FreeRTOSConfig.h
/*
 * FreeRTOS V202012.00
 * Copyright (C) 2020 Amazon.com, Inc. or its affiliates.  All Rights Reserved.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 *
 * http://www.FreeRTOS.org
 * http://aws.amazon.com/freertos
 *
 */
#ifndef FREERTOS_CONFIG_H
#define FREERTOS_CONFIG_H

/*-----------------------------------------------------------
 * Application specific definitions.
 *
 * These definitions should be adjusted for your particular hardware and
 * application requirements.
 *
 * THESE PARAMETERS ARE DESCRIBED WITHIN THE 'CONFIGURATION' SECTION OF THE
 * FreeRTOS API DOCUMENTATION AVAILABLE ON THE FreeRTOS.org WEB SITE.  See
 * http://www.freertos.org/a00110.html
 *----------------------------------------------------------*/

/*-----------------------------------------------------------
 * EXAMPLE FREERTOS CONFIG FILE
 * This config example is specifically made to support the POSIX and Windows FreeRTOS simulators. The following options
 * are NOT FreeRTOS options, and are instead provided by this module to make getting up and running with FreeRTOS easier
 * by providing default implementations for some of the functions.
 *----------------------------------------------------------*/
#include <assert.h>
/** Provide the recommended FreeRTOS default implementation for vApplicationGetIdleTaskMemory */
#define IMPLEMENT_DEFAULT_vApplicationGetIdleTaskMemory 1
/** Provide the recommended FreeRTOS default implementation for vApplicationGetTimerTaskMemory */
#define IMPLEMENT_DEFAULT_vApplicationGetTimerTaskMemory 1

/* Setting configSUPPORT_STATIC_ALLOCATION to 1 allows RTOS objects to be
created using only application supplied memory.  No dynamic memory allocation
will be performed. */
#define configSUPPORT_STATIC_ALLOCATION 1

/* Setting configSUPPORT_DYNAMIC_ALLOCATION to 0 results in all calls to
pvPortMalloc() returning NULL, and all calls to vPortFree() being ignored.
Therefore the application can be built without providing an implementation of
either of these functions (so none of the normal heap_n.c files described on
http://www.freertos.org/a00111.html are required).  Note that
configTOTAL_HEAP_SIZE is not defined. */
#define configSUPPORT_DYNAMIC_ALLOCATION 1

/* Other constants as described on http://www.freertos.org/a00110.html */
#define configUSE_PREEMPTION                    1
#define configUSE_PORT_OPTIMISED_TASK_SELECTION 0
#define configUSE_IDLE_HOOK                     0
#define configUSE_TICK_HOOK                     0
#define configUSE_DAEMON_TASK_STARTUP_HOOK      0
/* In this non-real time simulated environment the tick frequency has to be at least a multiple of the Win32
 *  tick frequency, and therefore very slow.
 */
#define configTICK_RATE_HZ             (1000)
#define configMINIMAL_STACK_SIZE       ((unsigned short)128)
#define configMAX_TASK_NAME_LEN        (15)
#define configUSE_TRACE_FACILITY       1
#define configUSE_16_BIT_TICKS         0
#define configIDLE_SHOULD_YIELD        1
#define configUSE_MUTEXES              1
#define configCHECK_FOR_STACK_OVERFLOW 0
#define configUSE_RECURSIVE_MUTEXES    1
#define configQUEUE_REGISTRY_SIZE      20
#define configUSE_MALLOC_FAILED_HOOK   0
#define configUSE_APPLICATION_TASK_TAG 1
#define configUSE_COUNTING_SEMAPHORES  1
#define configUSE_ALTERNATIVE_API      0
#define configUSE_QUEUE_SETS           1
#define configUSE_TASK_NOTIFICATIONS   1

/* Software timer related configuration options. */
#define configUSE_TIMERS             1
#define configTIMER_TASK_PRIORITY    (configMAX_PRIORITIES - 1)
#define configTIMER_QUEUE_LENGTH     20
#define configTIMER_TASK_STACK_DEPTH (configMINIMAL_STACK_SIZE * 2)

#define configMAX_PRIORITIES (7)

/* Run time stats gathering configuration options. */
#define configGENERATE_RUN_TIME_STATS 0
#define portCONFIGURE_TIMER_FOR_RUN_TIME_STATS()
/** The POSIX port defines this macro, the Windows port does not. */
#if !__unix__
  #define portGET_RUN_TIME_COUNTER_VALUE()
#endif

/* Co-routine related configuration options. */
#define configUSE_CO_ROUTINES 0

/* This demo makes use of one or more example stats formatting functions.  These
format the raw data provided by the uxTaskGetSystemState() function in to human
readable ASCII form.  See the notes in the implementation of vTaskList() within
FreeRTOS/Source/tasks.c for limitations. */
#define configUSE_STATS_FORMATTING_FUNCTIONS 0

/* Set the following definitions to 1 to include the API function, or zero
to exclude the API function.  In most cases the linker will remove unused
functions anyway. */
#define INCLUDE_vTaskPrioritySet               1
#define INCLUDE_uxTaskPriorityGet              1
#define INCLUDE_vTaskDelete                    1
#define INCLUDE_vTaskCleanUpResources          0
#define INCLUDE_vTaskSuspend                   1
#define INCLUDE_vTaskDelayUntil                1
#define INCLUDE_vTaskDelay                     1
#define INCLUDE_uxTaskGetStackHighWaterMark    1
#define INCLUDE_xTaskGetSchedulerState         1
#define INCLUDE_xTimerGetTimerDaemonTaskHandle 1
#define INCLUDE_xTaskGetIdleTaskHandle         1
#define INCLUDE_xTaskGetHandle                 1
#define INCLUDE_eTaskGetState                  1
#define INCLUDE_xSemaphoreGetMutexHolder       1
#define INCLUDE_xTimerPendFunctionCall         1
#define INCLUDE_xTaskAbortDelay                1

/* It is a good idea to define configASSERT() while developing.  configASSERT()
uses the same semantics as the standard C assert() macro. */
#define configASSERT(expr) assert(expr)

#endif /* FREERTOS_CONFIG_H */


tools/cmake/modules/FreeRTOSIPConfig.h
/*
 * FreeRTOS V202012.00
 * Copyright (C) 2020 Amazon.com, Inc. or its affiliates.  All Rights Reserved.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 *
 * https://www.FreeRTOS.org
 * https://github.com/FreeRTOS
 *
 */

/*****************************************************************************
 *
 * See the following URL for configuration information.
 * http://www.freertos.org/FreeRTOS-Plus/FreeRTOS_Plus_TCP/TCP_IP_Configuration.html
 *
 *****************************************************************************/

#ifndef FREERTOS_IP_CONFIG_H
#define FREERTOS_IP_CONFIG_H

#include <stdio.h>
#include <stdlib.h>

#include "FreeRTOS.h"

static inline UBaseType_t uxRand()
{
  return rand();
}

static inline BaseType_t ApplicationGetRandomNumber(uint32_t* number)
{
  *number = rand();
  return pdTRUE;
}
#define xApplicationGetRandomNumber(expr)      ApplicationGetRandomNumber(expr)
#define mainHOST_NAME                          "RTOSDemo"
#define mainDEVICE_NICK_NAME                   "linux_demo"
#define ipconfigNUM_NETWORK_BUFFER_DESCRIPTORS (2)

// LOOPBACK
// sudo tcpdump -D | grep lo
#define configNETWORK_INTERFACE_TO_USE 2

// ifconfig lo
#define configNET_MASK3 (255)
#define configNET_MASK2 (0)
#define configNET_MASK1 (0)
#define configNET_MASK0 (0)

#define configMAC_ISR_SIMULATOR_PRIORITY (configMAX_PRIORITIES - 1)

/* Set to 1 to print out debug messages.  If ipconfigHAS_DEBUG_PRINTF is set to
1 then FreeRTOS_debug_printf should be defined to the function used to print
out the debugging messages. */
#define ipconfigHAS_DEBUG_PRINTF 0
#if (ipconfigHAS_DEBUG_PRINTF == 1)
  #define FreeRTOS_debug_printf(X) printf X
#endif

/* Set to 1 to print out non debugging messages, for example the output of the
FreeRTOS_netstat() command, and ping replies.  If ipconfigHAS_PRINTF is set to 1
then FreeRTOS_printf should be set to the function used to print out the
messages. */
#define ipconfigHAS_PRINTF 0
#if (ipconfigHAS_PRINTF == 1)
  #define FreeRTOS_printf(X) vLoggingPrintf X
#endif

/* Define the byte order of the target MCU (the MCU FreeRTOS+TCP is executing
on).  Valid options are pdFREERTOS_BIG_ENDIAN and pdFREERTOS_LITTLE_ENDIAN. */
#define ipconfigBYTE_ORDER pdFREERTOS_LITTLE_ENDIAN

/* If the network card/driver includes checksum offloading (IP/TCP/UDP checksums)
then set ipconfigDRIVER_INCLUDED_RX_IP_CHECKSUM to 1 to prevent the software
stack repeating the checksum calculations. */
#define ipconfigDRIVER_INCLUDED_RX_IP_CHECKSUM 1

/* Several API's will block until the result is known, or the action has been
performed, for example FreeRTOS_send() and FreeRTOS_recv().  The timeouts can be
set per socket, using setsockopt().  If not set, the times below will be
used as defaults. */
#define ipconfigSOCK_DEFAULT_RECEIVE_BLOCK_TIME (5000)
#define ipconfigSOCK_DEFAULT_SEND_BLOCK_TIME    (5000)

/* Include support for LLMNR: Link-local Multicast Name Resolution
(non-Microsoft) */
#define ipconfigUSE_LLMNR (1)

/* Include support for NBNS: NetBIOS Name Service (Microsoft) */
#define ipconfigUSE_NBNS (1)

/* Include support for DNS caching.  For TCP, having a small DNS cache is very
useful.  When a cache is present, ipconfigDNS_REQUEST_ATTEMPTS can be kept low
and also DNS may use small timeouts.  If a DNS reply comes in after the DNS
socket has been destroyed, the result will be stored into the cache.  The next
call to FreeRTOS_gethostbyname() will return immediately, without even creating
a socket. */
#define ipconfigUSE_DNS_CACHE         (1)
#define ipconfigDNS_CACHE_NAME_LENGTH (16)
#define ipconfigDNS_CACHE_ENTRIES     (4)
#define ipconfigDNS_REQUEST_ATTEMPTS  (2)

/* The IP stack executes it its own task (although any application task can make
use of its services through the published sockets API). ipconfigUDP_TASK_PRIORITY
sets the priority of the task that executes the IP stack.  The priority is a
standard FreeRTOS task priority so can take any value from 0 (the lowest
priority) to (configMAX_PRIORITIES - 1) (the highest priority).
configMAX_PRIORITIES is a standard FreeRTOS configuration parameter defined in
FreeRTOSConfig.h, not FreeRTOSIPConfig.h. Consideration needs to be given as to
the priority assigned to the task executing the IP stack relative to the
priority assigned to tasks that use the IP stack. */
#define ipconfigIP_TASK_PRIORITY (configMAX_PRIORITIES - 2)

/* The size, in words (not bytes), of the stack allocated to the FreeRTOS+TCP
task.  This setting is less important when the FreeRTOS Win32 simulator is used
as the Win32 simulator only stores a fixed amount of information on the task
stack.  FreeRTOS includes optional stack overflow detection, see:
http://www.freertos.org/Stacks-and-stack-overflow-checking.html */
#define ipconfigIP_TASK_STACK_SIZE_WORDS (configMINIMAL_STACK_SIZE * 5)

/* ipconfigRAND32() is called by the IP stack to generate random numbers for
things such as a DHCP transaction number or initial sequence number.  Random
number generation is performed via this macro to allow applications to use their
own random number generation method.  For example, it might be possible to
generate a random number by sampling noise on an analogue input. */
extern UBaseType_t uxRand();
#define ipconfigRAND32() uxRand()

/* If ipconfigUSE_NETWORK_EVENT_HOOK is set to 1 then FreeRTOS+TCP will call the
network event hook at the appropriate times.  If ipconfigUSE_NETWORK_EVENT_HOOK
is not set to 1 then the network event hook will never be called.  See
http://www.FreeRTOS.org/FreeRTOS-Plus/FreeRTOS_Plus_UDP/API/vApplicationIPNetworkEventHook.shtml
*/
#define ipconfigUSE_NETWORK_EVENT_HOOK 1

/* Sockets have a send block time attribute.  If FreeRTOS_sendto() is called but
a network buffer cannot be obtained then the calling task is held in the Blocked
state (so other tasks can continue to executed) until either a network buffer
becomes available or the send block time expires.  If the send block time expires
then the send operation is aborted.  The maximum allowable send block time is
capped to the value set by ipconfigMAX_SEND_BLOCK_TIME_TICKS.  Capping the
maximum allowable send block time prevents prevents a deadlock occurring when
all the network buffers are in use and the tasks that process (and subsequently
free) the network buffers are themselves blocked waiting for a network buffer.
ipconfigMAX_SEND_BLOCK_TIME_TICKS is specified in RTOS ticks.  A time in
milliseconds can be converted to a time in ticks by dividing the time in
milliseconds by portTICK_PERIOD_MS. */
#define ipconfigUDP_MAX_SEND_BLOCK_TIME_TICKS (5000U / portTICK_PERIOD_MS)

/* If ipconfigUSE_DHCP is 1 then FreeRTOS+TCP will attempt to retrieve an IP
address, netmask, DNS server address and gateway address from a DHCP server.  If
ipconfigUSE_DHCP is 0 then FreeRTOS+TCP will use a static IP address.  The
stack will revert to using the static IP address even when ipconfigUSE_DHCP is
set to 1 if a valid configuration cannot be obtained from a DHCP server for any
reason.  The static configuration used is that passed into the stack by the
FreeRTOS_IPInit() function call. */
#define ipconfigUSE_DHCP 0

/* When ipconfigUSE_DHCP is set to 1, DHCP requests will be sent out at
increasing time intervals until either a reply is received from a DHCP server
and accepted, or the interval between transmissions reaches
ipconfigMAXIMUM_DISCOVER_TX_PERIOD.  The IP stack will revert to using the
static IP address passed as a parameter to FreeRTOS_IPInit() if the
re-transmission time interval reaches ipconfigMAXIMUM_DISCOVER_TX_PERIOD without
a DHCP reply being received. */
#define ipconfigMAXIMUM_DISCOVER_TX_PERIOD (120000U / portTICK_PERIOD_MS)

/* The ARP cache is a table that maps IP addresses to MAC addresses.  The IP
stack can only send a UDP message to a remove IP address if it knowns the MAC
address associated with the IP address, or the MAC address of the router used to
contact the remote IP address.  When a UDP message is received from a remote IP
address the MAC address and IP address are added to the ARP cache.  When a UDP
message is sent to a remote IP address that does not already appear in the ARP
cache then the UDP message is replaced by a ARP message that solicits the
required MAC address information.  ipconfigARP_CACHE_ENTRIES defines the maximum
number of entries that can exist in the ARP table at any one time. */
#define ipconfigARP_CACHE_ENTRIES 6

/* ARP requests that do not result in an ARP response will be re-transmitted a
maximum of ipconfigMAX_ARP_RETRANSMISSIONS times before the ARP request is
aborted. */
#define ipconfigMAX_ARP_RETRANSMISSIONS (5)

/* ipconfigMAX_ARP_AGE defines the maximum time between an entry in the ARP
table being created or refreshed and the entry being removed because it is stale.
New ARP requests are sent for ARP cache entries that are nearing their maximum
age.  ipconfigMAX_ARP_AGE is specified in tens of seconds, so a value of 150 is
equal to 1500 seconds (or 25 minutes). */
#define ipconfigMAX_ARP_AGE 150

/* Implementing FreeRTOS_inet_addr() necessitates the use of string handling
routines, which are relatively large.  To save code space the full
FreeRTOS_inet_addr() implementation is made optional, and a smaller and faster
alternative called FreeRTOS_inet_addr_quick() is provided.  FreeRTOS_inet_addr()
takes an IP in decimal dot format (for example, "192.168.0.1") as its parameter.
FreeRTOS_inet_addr_quick() takes an IP address as four separate numerical octets
(for example, 192, 168, 0, 1) as its parameters.  If
ipconfigINCLUDE_FULL_INET_ADDR is set to 1 then both FreeRTOS_inet_addr() and
FreeRTOS_indet_addr_quick() are available.  If ipconfigINCLUDE_FULL_INET_ADDR is
not set to 1 then only FreeRTOS_indet_addr_quick() is available. */
#define ipconfigINCLUDE_FULL_INET_ADDR 1

/* ipconfigNUM_NETWORK_BUFFER_DESCRIPTORS defines the total number of network buffer that
are available to the IP stack.  The total number of network buffers is limited
to ensure the total amount of RAM that can be consumed by the IP stack is capped
to a pre-determinable value. */
#define ipconfigNUM_NETWORK_BUFFER_DESCRIPTORS 60

/* A FreeRTOS queue is used to send events from application tasks to the IP
stack.  ipconfigEVENT_QUEUE_LENGTH sets the maximum number of events that can
be queued for processing at any one time.  The event queue must be a minimum of
5 greater than the total number of network buffers. */
#define ipconfigEVENT_QUEUE_LENGTH (ipconfigNUM_NETWORK_BUFFER_DESCRIPTORS + 5)

/* The address of a socket is the combination of its IP address and its port
number.  FreeRTOS_bind() is used to manually allocate a port number to a socket
(to 'bind' the socket to a port), but manual binding is not normally necessary
for client sockets (those sockets that initiate outgoing connections rather than
wait for incoming connections on a known port number).  If
ipconfigALLOW_SOCKET_SEND_WITHOUT_BIND is set to 1 then calling
FreeRTOS_sendto() on a socket that has not yet been bound will result in the IP
stack automatically binding the socket to a port number from the range
socketAUTO_PORT_ALLOCATION_START_NUMBER to 0xffff.  If
ipconfigALLOW_SOCKET_SEND_WITHOUT_BIND is set to 0 then calling FreeRTOS_sendto()
on a socket that has not yet been bound will result in the send operation being
aborted. */
#define ipconfigALLOW_SOCKET_SEND_WITHOUT_BIND 1

/* Defines the Time To Live (TTL) values used in outgoing UDP packets. */
#define ipconfigUDP_TIME_TO_LIVE 128
#define ipconfigTCP_TIME_TO_LIVE 128 /* also defined in FreeRTOSIPConfigDefaults.h */

/* USE_TCP: Use TCP and all its features */
#define ipconfigUSE_TCP (1)

/* USE_WIN: Let TCP use windowing mechanism. */
#define ipconfigUSE_TCP_WIN (1)

/* The MTU is the maximum number of bytes the payload of a network frame can
contain.  For normal Ethernet V2 frames the maximum MTU is 1500.  Setting a
lower value can save RAM, depending on the buffer management scheme used.  If
ipconfigCAN_FRAGMENT_OUTGOING_PACKETS is 1 then (ipconfigNETWORK_MTU - 28) must
be divisible by 8. */
#define ipconfigNETWORK_MTU 1200U

/* Set ipconfigUSE_DNS to 1 to include a basic DNS client/resolver.  DNS is used
through the FreeRTOS_gethostbyname() API function. */
#define ipconfigUSE_DNS 1

/* If ipconfigREPLY_TO_INCOMING_PINGS is set to 1 then the IP stack will
generate replies to incoming ICMP echo (ping) requests. */
#define ipconfigREPLY_TO_INCOMING_PINGS 1

/* If ipconfigSUPPORT_OUTGOING_PINGS is set to 1 then the
FreeRTOS_SendPingRequest() API function is available. */
#define ipconfigSUPPORT_OUTGOING_PINGS 0

/* If ipconfigSUPPORT_SELECT_FUNCTION is set to 1 then the FreeRTOS_select()
(and associated) API function is available. */
#define ipconfigSUPPORT_SELECT_FUNCTION 1

/* If ipconfigFILTER_OUT_NON_ETHERNET_II_FRAMES is set to 1 then Ethernet frames
that are not in Ethernet II format will be dropped.  This option is included for
potential future IP stack developments. */
#define ipconfigFILTER_OUT_NON_ETHERNET_II_FRAMES 1

/* If ipconfigETHERNET_DRIVER_FILTERS_FRAME_TYPES is set to 1 then it is the
responsibility of the Ethernet interface to filter out packets that are of no
interest.  If the Ethernet interface does not implement this functionality, then
set ipconfigETHERNET_DRIVER_FILTERS_FRAME_TYPES to 0 to have the IP stack
perform the filtering instead (it is much less efficient for the stack to do it
because the packet will already have been passed into the stack).  If the
Ethernet driver does all the necessary filtering in hardware then software
filtering can be removed by using a value other than 1 or 0. */
#define ipconfigETHERNET_DRIVER_FILTERS_FRAME_TYPES 1

/* The Linux simulator cannot really simulate MAC interrupts, and needs to
block occasionally to allow other tasks to run. */
#define configWINDOWS_MAC_INTERRUPT_SIMULATOR_DELAY (20 / portTICK_PERIOD_MS)

/* Advanced only: in order to access 32-bit fields in the IP packets with
32-bit memory instructions, all packets will be stored 32-bit-aligned, plus 16-bits.
This has to do with the contents of the IP-packets: all 32-bit fields are
32-bit-aligned, plus 16-bit(!) */
#define ipconfigPACKET_FILLER_SIZE 2U

/* Define the size of the pool of TCP window descriptors.  On the average, each
TCP socket will use up to 2 x 6 descriptors, meaning that it can have 2 x 6
outstanding packets (for Rx and Tx).  When using up to 10 TP sockets
simultaneously, one could define TCP_WIN_SEG_COUNT as 120. */
#define ipconfigTCP_WIN_SEG_COUNT 240

/* Each TCP socket has a circular buffers for Rx and Tx, which have a fixed
maximum size.  Define the size of Rx buffer for TCP sockets. */
#define ipconfigTCP_RX_BUFFER_LENGTH (1000)

/* Define the size of Tx buffer for TCP sockets. */
#define ipconfigTCP_TX_BUFFER_LENGTH (1000)

/* When using call-back handlers, the driver may check if the handler points to
real program memory (RAM or flash) or just has a random non-zero value. */
#define ipconfigIS_VALID_PROG_ADDRESS(x) ((x) != NULL)

/* Include support for TCP hang protection.  All sockets in a connecting or
disconnecting stage will timeout after a period of non-activity. */
#define ipconfigTCP_HANG_PROTECTION      (1)
#define ipconfigTCP_HANG_PROTECTION_TIME (30)

/* Include support for TCP keep-alive messages. */
#define ipconfigTCP_KEEP_ALIVE          (1)
#define ipconfigTCP_KEEP_ALIVE_INTERVAL (20) /* in seconds */

/* Required for 64-bit platforms */
#define ipconfigBUFFER_PADDING 14U

#define portINLINE __inline

#endif /* FREERTOS_IP_CONFIG_H */
